### ISSUE TITLE
Refactor import to comply with PEP 8

### ISSUE THREAD

**Issue by maximilianwerk** [jina-ai/jina#373]
**Describe the feature**
Currently jina uses relative imports everywhere. Furthermore, there are a lot of import on function level and not at the top level of the files. jina guidelines say, PEP 8 should be followed. Anyhow, [PEP 8 is explicit about imports](https://www.python.org/dev/peps/pep-0008/#imports): They should be put on top of each file and absolute imports are preferred above relative imports. Doing this makes later refactoring and moving of modules less error prone. This might be especially useful, since jina is at the beginning of its lifetime and thus might undergo quite some refactoring in the future.

In case, this is wanted, I am happy to do the refactoring.

**Your proposal**
Follow best practices and refactor the relative imports to absolute imports and move imports to the top of each file. As an example, the `jina/main/__init__.py` file would look like:

```
__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
__license__ = "Apache-2.0"

import sys

from pkg_resources import resource_filename

from jina.helper import colored
from jina.logging import default_logger
from jina.main import api
from jina.main.autocomplete import ac_table
from jina.main.parser import get_main_parser


def _get_run_args(print_args: bool = True):
    parser = get_main_parser()
    if len(sys.argv) > 1:
        args = parser.parse_args()
        default_args = parser.parse_args([sys.argv[1]])
        if print_args:
            with open(
                resource_filename("jina", "/".join(("resources", "jina.logo")))
            ) as fp:
                logo_str = fp.read()
            param_str = []
            for k, v in sorted(vars(args).items()):
                j = f'{k.replace("_", "-"): >30.30} = {str(v):30.30}'
                if getattr(default_args, k) == v:
                    param_str.append("   " + j)
                else:
                    param_str.append("🔧️ " + colored(j, "blue", "on_yellow"))
            param_str = "\n".join(param_str)
            default_logger.info(
                f'\n{logo_str}\n▶️  {" ".join(sys.argv)}\n{param_str}\n'
            )
        return args
    else:
        parser.print_help()
        exit()


def _quick_ac_lookup():
    if len(sys.argv) > 1:
        if sys.argv[1] == "commands":
            for k in ac_table["commands"]:
                print(k)
            exit()
        elif sys.argv[1] == "completions":
            if sys.argv[2] in ac_table["completions"]:
                for k in ac_table["completions"][sys.argv[2]]:
                    if k not in sys.argv:
                        print(k)
            exit()


def main():
    """The main entrypoint of the CLI """
    _quick_ac_lookup()
    args = _get_run_args()
    getattr(api, args.cli.replace("-", "_"))(args)

```

**Environment**
```
jina                          0.1.7
jina-proto                    0.0.20
jina-vcs-tag                  (unset)
libzmq                        4.3.2
pyzmq                         1.18.3
protobuf                      3.11.3
proto-backend                 cpp
grpcio                        1.28.1
ruamel.yaml                   0.16.10
python                        3.8.0
platform                      Linux
platform-release              5.3.0-51-generic
platform-version              #44-Ubuntu SMP Wed Apr 22 21:09:44 UTC 2020
architecture                  x86_64
processor                     x86_64
jina-resources                XXX
JINA_ARRAY_QUANT              (unset)
JINA_CONTRIB_MODULE           (unset)
JINA_CONTRIB_MODULE_IS_LOADING(unset)
JINA_CONTROL_PORT             (unset)
JINA_DEFAULT_HOST             (unset)
JINA_EXECUTOR_WORKDIR         (unset)
JINA_FULL_CLI                 (unset)
JINA_IPC_SOCK_TMP             (unset)
JINA_LOG_FILE                 (unset)
JINA_LOG_LONG                 (unset)
JINA_LOG_NO_COLOR             (unset)
JINA_LOG_PROFILING            (unset)
JINA_LOG_SSE                  (unset)
JINA_LOG_VERBOSITY            (unset)
JINA_PROFILING                (unset)
JINA_SOCKET_HWM               (unset)
JINA_STACK_CONFIG             (unset)
JINA_TEST_CONTAINER           (unset)
JINA_TEST_PRETRAINED          (unset)
JINA_VCS_VERSION              (unset)
JINA_VERSION                  (unset)
JINA_WARN_UNNAMED             (unset)
```

**Comment by hanxiao** [comment]
We use relative import everywhere (in the header), as the absolute import is very verbose to me. Especially when module A and module B are placed very deep in the package layout and then module A needs to import module B. When using relative import, a simple `from .B import BClass` is sufficient, saving us from over-lengthy full path. In Jina package design, such deep layout is pretty common.

Besides, https://www.python.org/dev/peps/pep-0008/#imports and https://stackoverflow.com/a/16748366 suggest that relative imports is acceptable as long as the intention is to make the code more clear.

I think it would be fine as long as we use relative import everywhere and keep the style consistent. I will wait for more community feedback on this thread before moving on.

**Comment by maximilianwerk** [comment]
Fair enough. Thanks for the clarification.

What about moving imports to the top of files? Do you also have good arguments, why they should stay inside functions?

**Comment by hanxiao** [comment]
Glad you mention that. I believe many people may feel strange when they see some imports are deeply buried in the code. Why not put them all in the header?

The reason is tri-partite: 
- for the sake of speed
- to avoid cyclic imports
- reduce module not found error

1. **Speed**

One good example is https://github.com/jina-ai/jina/blob/master/jina/main/parser.py  where I put all imports deep in the functions not in the header. The reason is that we want `jina` command to run fast in the terminal. When we type `jina` or simple command such as `jina -v`, the program should **not** imports every modules into memory, it should only look for the module/class that are really needed. Let's be honest, Python is not the fastest programming language when it comes to the execution speed. We care about speed and smooth experience.

Besides, **importing is implemented nontrivially in Jina.** We have some metaclass defined for `BaseExecutor`, `BaseDriver` and `Flow` which inject couple of black magic (YAML serialization, meta tags, etc.) while importing. There are lots of things going on when importing an Executor (for the first time).

2. **Cyclic imports**

Jina has a nontrivial package layout. Cyclic module dependencies have been well-thought and solved with the greatest effort, yet they can not be solved **completely**. The best way to avoid cyclic dependencies is to only import when you need them. Not at the header.

3. **Reduce module not found error**

Despite the nature of being universal search engine, Jina has only five base dependencies: numpy, protobuf, grpc, ruamel.yaml and pyzmq. All other packages, dependencies are optional and considered as "plugins". 

Let's say we use `flask` for exposing the endpoints to dashboard, but this feature is completely optional, if you don't use dashboard at all, then we don't ask you to install flask. 

In another case, people who use Tensorflow-Transformer don't need Pytorch, such autonomy is leaving to the AI engineer. In Jina we need to ensure our package can be run without all those extra dependencies. 

You can look at many subclass inherited from `BaseEncoder`, check how we put `import` under `post_init()`, and here is why https://docs.jina.ai/chapters/extend/executor.html#override-post-init

Another article you may want to read when it comes to plugin-dependencies in Python: https://hanxiao.io/2019/11/07/A-Better-Practice-for-Managing-extras-require-Dependencies-in-Python/

In summary, "late-import", or "lazy-import", "import-as-you-go" how ever you call it, is considered as the **best practice in Jina** given our import mechanism and dependencies layout.
